src/main/kotlin/io/sip3/salto/ce/sip/SipTransactionHandler.kt

```diff
@@ -0,0 +1,178 @@
+import io.sip3.commons.util.format
+import io.sip3.salto.ce.Attributes
+import io.sip3.salto.ce.RoutesCE
+import io.sip3.salto.ce.USE_LOCAL_CODEC
+import io.sip3.salto.ce.domain.Packet
+import io.sip3.salto.ce.util.transactionId
+import io.vertx.core.AbstractVerticle
+import io.vertx.core.json.JsonObject
+import mu.KotlinLogging
+import java.time.format.DateTimeFormatter
+import kotlin.math.abs
+
+/**
+ * Handles SIP Transactions
+ */
+open class SipTransactionHandler : AbstractVerticle() {
+
+    private val logger = KotlinLogging.logger {}
+
+    companion object {
+
+        val PREFIX = RoutesCE.sip + "_transaction"
+    }
+
+    private var timeSuffix: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")
+    private var expirationDelay: Long = 1000
+    private var aggregationTimeout: Long = 60000
+    private var terminationTimeout: Long = 5000
+
+    private var recordCallUsersAttributes = false
+    private var instances = 1
+
+    private var transactions = mutableMapOf<String, SipTransaction>()
+
+    override fun start() {
+        config().getString("time-suffix")?.let {
+            timeSuffix = DateTimeFormatter.ofPattern(it)
+        }
+        config().getJsonObject("sip")?.getJsonObject("transaction")?.let { config ->
+            config.getLong("expiration-delay")?.let {
+                expirationDelay = it
+            }
+            config.getLong("aggregation-timeout")?.let {
+                aggregationTimeout = it
+            }
+            config.getLong("termination-timeout")?.let {
+                terminationTimeout = it
+            }
+        }
+        config().getJsonObject("attributes")?.getBoolean("record-call-users")?.let {
+            recordCallUsersAttributes = it
+        }
+        config().getJsonObject("vertx")?.getInteger("instances")?.let {
+            instances = it
+        }
+
+        vertx.setPeriodic(expirationDelay) {
+            terminateExpiredTransactions()
+        }
+
+        val index = config().getInteger("index")
+        vertx.eventBus().localConsumer<Pair<Packet, SIPMessage>>(PREFIX + "_$index") { event ->
+            try {
+                val (packet, message) = event.body()
+                handle(packet, message)
+            } catch (e: Exception) {
+                logger.error("SipTransactionHandler 'handle()' failed.", e)
+            }
+        }
+    }
+
+    open fun handle(packet: Packet, message: SIPMessage) {
+        transactions.getOrPut(message.transactionId()) { SipTransaction() }.apply {
+            addMessage(packet, message)
+        }
+    }
+
+    open fun terminateExpiredTransactions() {
+        val now = System.currentTimeMillis()
+
+        transactions.filterValues { transaction ->
+            (transaction.terminatedAt?.let { it + terminationTimeout } ?: transaction.createdAt + aggregationTimeout) < now
+        }.forEach { (tid, transaction) ->
+            transactions.remove(tid)
+            writeAttributes(transaction)
```

Comment:
Here we need to route transaction to SipXXXHandlers instead of trying to `writeAttributes` and `writeToDatabase`. Imagine if you will write to database every REGISTER transaction. That's not needed for sure. Let's better route from here to SipCallHandler, SipRegisterHandler, SipOptionsHandler.... Let if better be copy/paste but it will be clean and simple logic...